### PR TITLE
Support of custom URL path for alert

### PR DESCRIPTION
**Pull request by kreedom**
New property `external.alert.source` has been introduced. User can overwrite the default path for getting alert link and send it as source URL to Alertmanager
Address https://github.com/VictoriaMetrics/VictoriaMetrics/issues/517

### CODE REVIEW

File: app/vmalert/notifier/template_func.go [hunk 1]
```diff
@@ -178,3 +180,27 @@ const second = int64(time.Second / minimumTick)
 func (t Time) Time() time.Time {
 	return time.Unix(int64(t)/second, (int64(t)%second)*nanosPerTick)
 }
+
+// AlertSourceFunc returns alert generator url if alert source template is specified
+func AlertSourceFunc(baseURL *url.URL, sourceTemplate string, validateTemplate bool) (AlertURLGenerator, error) {
```
Comment (hagen1778) [inline app/vmalert/notifier/template_func.go:185]: I see no need of having this function in `notifier` package

File: app/vmalert/main.go [hunk 2]
```diff
@@ -166,6 +172,19 @@ func getExternalURL(externalURL, httpListenAddr string, isSecure bool) (*url.URL
...
+	if err != nil {
+		return nil, err
+	}
+	if sourceURL != nil {
+		return sourceURL, nil
+	}
+	return func(alert notifier.Alert) string {
+		return fmt.Sprintf("%s/api/v1/%s/%s/status", externalURL, strconv.FormatUint(alert.GroupID, 10), strconv.FormatUint(alert.ID, 10))
+	}, nil
+}
```
Comment (hagen1778) [inline app/vmalert/main.go:186]:
```suggestion
func getAlertURLGenerator(externalURL *url.URL) (notifier.AlertURLGenerator, error) {
	if *externalAlertSource == "" {
		return func(alert notifier.Alert) string {
			return fmt.Sprintf("%s/api/v1/%s/%s/status", externalURL, strconv.FormatUint(alert.GroupID, 10), strconv.FormatUint(alert.ID, 10))
		}, nil
	}
	
	if *validateTemplates {
		if err := notifier.ValidateTemplates(map[string]string{
			"template": *externalAlertSource,
		}); err != nil {
			return nil, fmt.Errorf("error validating source template %s:%w", *externalAlertSource, err)
		}
	}
	m := map[string]string{
		"tpl": *externalAlertSource,
	}
	return func(alert notifier.Alert) string {
		templated, err := alert.ExecTemplate(m)
		if err != nil {
			logger.Errorf("can not exec source template %s", err)
		}
		return fmt.Sprintf("%s/%s", externalURL, templated["tpl"])
	}, nil
}
```

File: app/vmalert/main.go [hunk 3]
```diff
@@ -76,13 +80,15 @@ func main() {
 		logger.Fatalf("can not get external url: %s ", err)
 	}
 	notifier.InitTemplateFunc(eu)
```
Comment (hagen1778) [inline app/vmalert/main.go:82]: Should this be actioned after new URL was generated? I wonder how functions like `pathEscape` would behave with new flag set. Btw, can we cover it with tests as well?

File: app/vmalert/main.go [hunk 4]
```diff
@@ -76,13 +80,15 @@ func main() {
...
+	aug, err := getAlertURLGenerator(eu)
+	if err != nil {
+		logger.Fatalf("can not get alert url generator func: %s ", err)
```
[the remark this answers is not in the document]
Comment (kreedom) [reply app/vmalert/main.go:85]: context comes with downstream error

[the remark this answers is not in the document]
Comment (kreedom) [reply app/vmalert/main.go]: text is changed, context comes with downstream error

File: app/vmalert/main.go (hunk 3 above)
Comment (kreedom) [reply app/vmalert/main.go:80]: i added the test, i don't think we need to re-init template funcs

**Pull request merged**